crates/solvers/src/boundary/liquidity/limit_order.rs

```diff
@@ -0,0 +1,192 @@
+use {
+    crate::domain::liquidity::limit_order::{LimitOrder, TakerAmount},
+    contracts::ethcontract::{H160, U256},
+    shared::baseline_solver::BaselineSolvable,
+};
+
+impl BaselineSolvable for LimitOrder {
+    fn get_amount_out(&self, out_token: H160, (in_amount, in_token): (U256, H160)) -> Option<U256> {
+        if in_token != self.taker.token.0
+            || out_token != self.maker.token.0
+            || in_amount > self.taker.amount
+        {
+            return None;
+        }
+
+        let fee_adjusted_amount = in_amount.checked_sub(self.fee.0)?;
+        if self.maker.amount > self.taker.amount {
```

Comment:
We do not operate with floating point numbers here. Consider an example where the ratio is 0.5. Then, we should multiply the amounts that could lead to U256 overflow.

Reply:
You could also get around that by not computing the ratio first and instead do all operations in order which is what we usually do:
```rust
fee_adjusted_amount * make.amount / taker.amount
```
If you want to be super careful in order to avoid overflows on the multiplication you could even do `a.full_mul(b)` which promotes the result to `U512`.
One thing to note throughout all this is that we should try to mimic what the actual smart contract does as closely as possible. Imagine they have an error in their equation or even some rounding issues. Then it does us no good if we have a theoretically better implementation if a solution generated by using it ends up reverting.

Reply:
Also let's add a small comment in the code explaining this non trivial rationale...

Reply:
True, tried to follow the contract's implementation: https://github.com/0xProject/protocol/blob/e66307ba319e8c3e2a456767403298b576abc85e/contracts/zero-ex/contracts/src/features/native_orders/NativeOrdersSettlement.sol#L469-L509